drivers/can/can_rcar.c

```diff
@@ -0,0 +1,1042 @@
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	uint8_t tcr = 0;
+	int ret = 0;
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+	switch (mode) {
+	case CAN_NORMAL_MODE:
+		tcr = 0;
+		break;
+	/*Controller is not allowed to send dominant bits*/
+	case CAN_SILENT_MODE:
+		tcr = RCAR_CAN_TCR_LISTEN_ONLY | RCAR_CAN_TCR_TSTE;
+		break;
+	/*Controller is in loopback mode (receive own messages)*/
+	case CAN_LOOPBACK_MODE:
+		tcr = RCAR_CAN_TCR_INT_LOOP | RCAR_CAN_TCR_TSTE;
+		break;
+	/*Combination of loopback and silent*/
+	case CAN_SILENT_LOOPBACK_MODE:
+		ret = -ENOSYS;
+		goto unlock;
+	}
+
+	/* Writing to TCR registers must be done in halt mode */
+	ret = can_rcar_enter_halt_mode(config);
+	if (ret) {
+		goto unlock;
+	}
+
+	sys_write8(tcr, config->reg_addr + RCAR_CAN_TCR);
+	/* Go back to operation mode */
+	ret = can_rcar_enter_operation_mode(config);
+
+unlock:
+	k_mutex_unlock(&data->inst_mutex);
+	return ret;
+}
+
+/* Bit Configuration Register settings */
+#define RCAR_CAN_BCR_TSEG1(x)   (((x) & 0x0f) << 20)
+#define RCAR_CAN_BCR_BPR(x)     (((x) & 0x3ff) << 8)
+#define RCAR_CAN_BCR_SJW(x)     (((x) & 0x3) << 4)
+#define RCAR_CAN_BCR_TSEG2(x)   ((x) & 0x07)
+
+static void can_rcar_set_bittiming(const struct can_rcar_cfg *config,
+				   const struct can_timing *timing)
+{
+	uint32_t bcr;
+
+	bcr = RCAR_CAN_BCR_TSEG1(timing->phase_seg1 + timing->prop_seg - 1) |
+	      RCAR_CAN_BCR_BPR(timing->prescaler - 1) |
+	      RCAR_CAN_BCR_SJW(timing->sjw - 1) |
+	      RCAR_CAN_BCR_TSEG2(timing->phase_seg2 - 1);
+
+	/* Don't overwrite CLKR with 32-bit BCR access; CLKR has 8-bit access.
+	 * All the registers are big-endian but they get byte-swapped on 32-bit
+	 * read/write (but not on 8-bit, contrary to the manuals)...
+	 */
+	sys_write32((bcr << 8) | RCAR_CAN_CLKR_CLKP2,
+		    config->reg_addr + RCAR_CAN_BCR);
+}
+
+int can_rcar_set_timing(const struct device *dev,
+			const struct can_timing *timing,
+			const struct can_timing *timing_data)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	int ret = 0;
+
+	ARG_UNUSED(timing_data);
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+
+	/* Changing bittiming should be done in reset mode */
+	ret = can_rcar_enter_reset_mode(config, true);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	can_rcar_set_bittiming(config, timing);
+
+	/* Go back to operation mode */
+	ret = can_rcar_enter_operation_mode(config);
+
+unlock:
+	k_mutex_unlock(&data->inst_mutex);
+	return ret;
+}
+
+static void can_rcar_register_state_change_isr(const struct device *dev,
+					       can_state_change_isr_t isr)
+{
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+
+	data->state_change_isr = isr;
+}
+
+static enum can_state can_rcar_get_state(const struct device *dev,
+					 struct can_bus_err_cnt *err_cnt)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+
+	if (err_cnt) {
```

Comment:
bool cast ? is it authorized on the type ?

Reply:
fixed